src/EventStore.Core/TransactionLog/Scavenging/Stages/ChunkExecutor.cs

```diff
@@ -0,0 +1,375 @@
+
+			var scavengePos = _chunkSize * startFromChunk;
+			var upTo = scavengePoint.Position;
+			while (scavengePos < upTo) {
+				// in bounds because we stop before the scavenge point
+				var physicalChunk = _chunkManager.GetChunkReaderFor(scavengePos);
+
+				if (!physicalChunk.IsReadOnly)
+					throw new Exception(
+						$"Reached open chunk before scavenge point. " +
+						$"Chunk {physicalChunk.ChunkStartNumber}. ScavengePoint: {upTo}.");
+
+				yield return physicalChunk;
+
+				scavengePos = physicalChunk.ChunkEndPosition;
+			}
+		}
+
+		private void ExecutePhysicalChunk(
+			float physicalWeight,
+			ScavengePoint scavengePoint,
+			IScavengeStateForChunkExecutorWorker<TStreamId> state,
+			ITFChunkScavengerLog scavengerLogger,
+			IChunkReaderForExecutor<TStreamId, TRecord> sourceChunk,
+			Stopwatch sw,
+			CancellationToken cancellationToken) {
+
+			sw.Restart();
+
+			int chunkStartNumber = sourceChunk.ChunkStartNumber;
+			long chunkStartPos = sourceChunk.ChunkStartPosition;
+			int chunkEndNumber = sourceChunk.ChunkEndNumber;
+			long chunkEndPos = sourceChunk.ChunkEndPosition;
+			var oldChunkName = sourceChunk.Name;
+
+			Log.Trace(
+				"SCAVENGING: started to scavenge physical chunk: {oldChunkName} " +
+				"with weight {physicalWeight:N0}. " +
+				"{chunkStartNumber} => {chunkEndNumber} ({chunkStartPosition} => {chunkEndPosition})",
+				oldChunkName,
+				physicalWeight,
+				chunkStartNumber, chunkEndNumber, chunkStartPos, chunkEndPos);
+
+			IChunkWriterForExecutor<TStreamId, TRecord> outputChunk;
+			try {
+				outputChunk = _chunkManager.CreateChunkWriter(sourceChunk);
+				Log.Trace(
+					"SCAVENGING: Resulting temp chunk file: {tmpChunkPath}.", 
+					Path.GetFileName(outputChunk.FileName));
+
+			} catch (IOException ex) {
+				Log.ErrorException(ex,
+					"IOException during creating new chunk for scavenging purposes. " +
+					"Stopping scavenging process...");
+				throw;
+			}
+
+			try {
+				var cancellationCheckCounter = 0;
+				var discardedCount = 0;
+				var keptCount = 0;
+
+				// nonPrepareRecord and prepareRecord ae reused through the iteration
+				var nonPrepareRecord = new RecordForExecutor<TStreamId, TRecord>.NonPrepare();
+				var prepareRecord = new RecordForExecutor<TStreamId, TRecord>.Prepare();
+
+				foreach (var isPrepare in sourceChunk.ReadInto(nonPrepareRecord, prepareRecord)) {
+					if (isPrepare) {
+						if (ShouldDiscard(state, scavengePoint, prepareRecord)) {
+							discardedCount++;
+						} else {
+							keptCount++;
+							outputChunk.WriteRecord(prepareRecord);
+						}
+					} else {
+						keptCount++;
+						outputChunk.WriteRecord(nonPrepareRecord);
+					}
+
+					if (++cancellationCheckCounter == _cancellationCheckPeriod) {
+						cancellationCheckCounter = 0;
+						cancellationToken.ThrowIfCancellationRequested();
+					}
+				}
+
+				Log.Trace(
+					"SCAVENGING: Scavenging {oldChunkName} traversed {recordsCount} including {filteredCount}.",
```

Comment:
suggestion: logging the number of records discarded may be more intuitive/useful than the number included

Reply:
done :+1:

Reply:
thanks!